### PR TITLE
Add and use VersionedDeprecationWarning

### PR DESCRIPTION
**Pull request by irenedea**
We should include versions with our deprecation warnings so we can easily tell when we should officially remove something.

When adding a new VersionedDeprecationWarning, one should pass the most recent released foundry version pus two versions to the `remove_version` argument.

### CODE REVIEW

File: llmfoundry/utils/warnings.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+
+class VersionedDeprecationWarning(DeprecationWarning):
```
Comment (dakinggg) [inline llmfoundry/utils/warnings.py:5]: Should we upstream this to composer?

File: llmfoundry/utils/warnings.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    """A custom deprecation warning class that includes version information.
+
+    Attributes:
+        message (str): The deprecation message describing why the feature is deprecated.
+        after_version (str): The version after which the feature will be deprecated.
```
Comment (dakinggg) [inline llmfoundry/utils/warnings.py:10]: The wording here isn't totally clear to me. Should it be "Feature X is deprecated and will be removed after version Y"?

File: llmfoundry/utils/warnings.py (hunk 1 above)
Comment (irenedea) [reply llmfoundry/utils/warnings.py:5]: hmm could but then we'd need to wait for a release to use it in foundry?

Comment (dakinggg) [reply llmfoundry/utils/warnings.py:5]: yeah you can also just do it in this pr and open a companion PR to composer

**Review by dakinggg — COMMENTED** [review]
I think the flash attention v1 deprecation is the only one you need to keep and the others can be deleted now.

File: llmfoundry/utils/warnings.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+
+class VersionedDeprecationWarning(DeprecationWarning):
+    """A custom deprecation warning class that includes version information.
+
+    Attributes:
+        message (str): The deprecation message describing why the feature is deprecated.
+        after_version (str): The version after which the feature will be removed.
```
Comment (dakinggg) [inline llmfoundry/utils/warnings.py:10]: update docstring

File: llmfoundry/callbacks/generate_callback.py [hunk 4]
```diff
@@ -20,10 +22,10 @@ def __init__(self, prompts: List[str], batch_log_interval: int,
                  **kwargs: Any):
 
         warnings.warn(
-            ('Accessing llmfoundry.callbacks.generate_callback.Generate '
-             'is deprecated and will be removed in a future release. '
-             'Please use composer.callbacks.Generate instead.'),
-            DeprecationWarning,
+            VersionedDeprecationWarning('Accessing llmfoundry.callbacks.generate_callback.Generate ' + \
+             'is deprecated and will be removed in a future release. Please use composer.callbacks.Generate instead.',
```
Comment (dakinggg) [inline llmfoundry/callbacks/generate_callback.py:26]:
```suggestion
             'is deprecated. Please use composer.callbacks.Generate instead.',
```

File: llmfoundry/data/packing.py [hunk 5]
```diff
@@ -433,9 +435,11 @@ def profile(raw_batch_size: int) -> Tuple[Optional[float], Optional[float]]:
     import warnings
 
     warnings.warn(
-        DeprecationWarning(
+        VersionedDeprecationWarning(
             'Please use scripts/misc/profile_packing.py to profile packing.' +
-            'This script will be removed in later releases.'))
+            'This script will be removed in later releases.',
```
Comment (dakinggg) [inline llmfoundry/data/packing.py:440]:
```suggestion
```

File: scripts/train/train.py [hunk 6]
```diff
@@ -219,8 +221,10 @@ def main(cfg: DictConfig) -> Trainer:
                                           default_value=None)
         if eval_gauntlet_config is not None:
             warnings.warn(
-                'Use of the key `model_gauntlet` is deprecated, please use the key `eval_gauntlet`',
-                DeprecationWarning)
+                VersionedDeprecationWarning(
+                    'Use of the key `model_gauntlet` is deprecated, please use the key `eval_gauntlet`',
```
Comment (dakinggg) [inline scripts/train/train.py:225]:
```suggestion
                    'Use of the key `model_gauntlet` is deprecated, please use the key `eval_gauntlet`.',
```

**Review by dakinggg — APPROVED** [review]
Just some tiny wording nits that don't really matter cause you're about to delete them anyway. Thanks!

File: llmfoundry/utils/warnings.py (hunk 3 above)
Comment (irenedea) [inline llmfoundry/utils/warnings.py:10]:
```suggestion
        remove_version (str): The version after which the feature will be removed.
```

**Pull request merged**